gensim/similarities/nmslib.py

```diff
@@ -0,0 +1,210 @@
+from gensim.models.keyedvectors import WordEmbeddingsKeyedVectors
+try:
+    import nmslib
+except ImportError:
+    raise ImportError(
+        "Nmslib has not been installed, if you wish to use the nmslib indexer, please run `pip install nmslib`"
+    )
+
+
+class NmslibIndexer(object):
+    """This class allows to use `Nmslib <https://github.com/nmslib/nmslib>`_ as indexer for `most_similar` method
+    from :class:`~gensim.models.word2vec.Word2Vec`, :class:`~gensim.models.doc2vec.Doc2Vec`,
+    :class:`~gensim.models.fasttext.FastText` and :class:`~gensim.models.keyedvectors.Word2VecKeyedVectors` classes.
+
+    """
+
+    def __init__(self, model=None, index_params=None, query_time_params=None):
+        """
+        Parameters
+        ----------
+        model : :class:`~gensim.models.base_any2vec.BaseWordEmbeddingsModel`, optional
+            Model, that will be used as source for index.
+        index_params : dict, optional
+            index_params for Nmslib indexer.
+        query_time_params : dict, optional
+            query_time_params for Nmslib indexer.
+
+        Examples
+        --------
+        .. sourcecode:: pycon
+
+            >>> from gensim.similarities.nmslib import NmslibIndexer
+            >>> from gensim.models import Word2Vec
+            >>>
+            >>> sentences = [['cute', 'cat', 'say', 'meow'], ['cute', 'dog', 'say', 'woof']]
+            >>> model = Word2Vec(sentences, min_count=1, seed=1)
+            >>>
+            >>> indexer = NmslibIndexer(model)
+            >>> model.most_similar("cat", topn=2, indexer=indexer)
+            [('cat', 1.0), ('meow', 0.5595494508743286)]
+
+        """
+        if index_params is None:
+            index_params = {'M': 10, 'indexThreadQty': 1, 'efConstruction': 100, 'post': 0}
+        if query_time_params is None:
+            query_time_params = {'efSearch': 100}
+
+        self.index = None
+        self.labels = None
+        self.model = model
+        self.index_params = index_params
+        self.query_time_params = query_time_params
+
+        if model:
+            if isinstance(self.model, Doc2Vec):
+                self.build_from_doc2vec()
+            elif isinstance(self.model, (Word2Vec, FastText)):
+                self.build_from_word2vec()
+            elif isinstance(self.model, (WordEmbeddingsKeyedVectors, KeyedVectors)):
+                self.build_from_keyedvectors()
+            else:
+                raise ValueError("Only a Word2Vec, Doc2Vec, FastText or KeyedVectors instance can be used")
+
+    def save(self, fname, protocol=2):
+        """Save NmslibIndexer instance.
+
+        Parameters
+        ----------
+        fname : str
+            Path to output file,
+            will produce 2 files: `fname` - parameters and `fname`.d - :class:`~nmslib.NmslibIndex`.
+        protocol : int, optional
+            Protocol for pickle.
+
+        Notes
+        -----
+        This method save **only** index (**model isn't preserved**).
```

Comment:
I think it'd be more helpful to the user to mention:

```suggestion
        This method saves **only** the index (**the model isn't preserved**).
        If you want to save the model, do that explicitly by calling the model.save method.
```

Reply:
Thanks. Changed.